Redis TTL

Talkabout opened the pull request:
This is my first time changing unbound code so sorry for any "obvious" errors :)

Currently unbound does not provide any TTL information to Redis when storing values. This makes setups quite complex as every single redis instance needs to take care of eviction policies, which might not always be trivial to define. The code does 4 things:

1. makes sure that cache backend receives the ttl extracted from the dns respone in "store"-function
2. checks whether it makes sense to set a ttl on redis records. This is the case if either "serve-expired" is set to "no" or there is a value for "serve-expired-ttl".
3. in case "serve-expired-ttl" is set, this value is added to the ttl received to make sure, that there is even a chance to provide an expired record in the configured time frame
4. if ttl is to be set the redis command is extended with " EX " + ttl

I am running the code for a few days now and there are no issues so far. Which, of course, does not mean there there aren't any :)

some things to discuss:

1. I have used "sizeof(uint64_t)" to calculate the string length. This might be wrong.
2. I am using "EX" command extension for redis, which is available only with redis 2.6.12. Reason is that I wanted to have one command being executed and not multiple. "SETEX" seems to be kind of deprecated with the addition of "EX". Question here is, which redis versions need to be supported?

Thanks!

Talkabout committed:
added logic for redis to honor ttl when serve_expired is not enabled

Talkabout committed:
Merge branch 'master' of https://github.com/Talkabout/unbound

Talkabout committed:
honor 'server_expired_ttl' in redis

gthess commented on cachedb/cachedb.c:
```diff
@@ -160,7 +160,7 @@ testframe_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
 static void
 testframe_store(struct module_env* env, struct cachedb_env* cachedb_env,
-	char* key, uint8_t* data, size_t data_len)
+	char* key, uint8_t* data, size_t data_len, uint64_t ttl)
```
- This could be `time_t`. `qstate->return_msg->rep->ttl` is already stored as `time_t` so no reason to cast here when calling.
- Because the attribute is unused in this function you should use `ATTR_UNUSED()` to wrap the ttl attribute. Otherwise if `-Wextra` is used when compiling, it will result in a warning.

gthess commented on cachedb/cachedb.c:
```diff
@@ -606,7 +606,8 @@ cachedb_extcache_store(struct module_qstate* qstate, struct cachedb_env* ie)
 		sldns_buffer_begin(qstate->env->scratch_buffer),
-		sldns_buffer_limit(qstate->env->scratch_buffer));
+		sldns_buffer_limit(qstate->env->scratch_buffer),
+		(uint64_t)qstate->return_msg->rep->ttl);
```
After the change to `time_t` this cast is no longer needed.

gthess commented on cachedb/cachedb.h:
```diff
@@ -84,7 +84,7 @@ struct cachedb_backend {
 	/** Store (env, cachedb_env, key, data, data_len) */
 	void (*store)(struct module_env*, struct cachedb_env*, char*,
-		uint8_t*, size_t);
+		uint8_t*, size_t, uint64_t);
```
Change to `time_t`.

gthess commented on cachedb/redis.c:
```diff
@@ -47,6 +47,7 @@
 #include "util/alloc.h"
 #include "util/config_file.h"
 #include "sldns/sbuffer.h"
+#include <stdbool.h>
```
Booleans may come with a few portability issues to other platforms. Use plain int instead.

gthess commented on cachedb/redis.c:
```diff
@@ -249,16 +250,37 @@ redis_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
-	char cmdbuf[4+(CACHEDB_HASHSIZE/8)*2+3+1]; /* "SET " + key + " %b" */
 	int n;
+	int size;
+	bool set_ttl = !env->cfg->serve_expired || env->cfg->serve_expired_ttl > 0;
```
Booleans may come with a few portability issues to other platforms. Use plain int instead.

gthess commented on cachedb/redis.c:
```diff
@@ -249,16 +250,37 @@ redis_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
 static void
 redis_store(struct module_env* env, struct cachedb_env* cachedb_env,
-	char* key, uint8_t* data, size_t data_len)
+	char* key, uint8_t* data, size_t data_len, uint64_t ttl)
```
Change to `time_t`.

gthess commented on cachedb/redis.c:
```diff
@@ -249,16 +250,37 @@ redis_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
+		size = 4+(CACHEDB_HASHSIZE/8)*2+3+1;
+	}
+	else {
+		size = 4+(CACHEDB_HASHSIZE/8)*2+3+4+sizeof(uint64_t)+1;
```
`sizeof()` will give the byte size of the type, not the byte size of the string representation.

Talkabout replied:
what is the best way to get the length of the time_t value here?

gthess replied:
The over-the-top way would be something like:
```
int ttl_string_size = snprintf(0, 0, "%u", (uint32_t) ttl);
// Check that the value returned is not negative and proceed
```
But this is too much for this. Given that later you will print a `uint32_t` with a max value of 4,294,967,295 I would just use 10 there.

gthess commented on cachedb/redis.c:
```diff
@@ -249,16 +250,37 @@ redis_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
+		n = snprintf(cmdbuf, sizeof(cmdbuf), "SET %s %%b", key);
+	}
+	else {
+		verbose(VERB_ALGO, "redis_store %s (%d bytes) with ttl %d", key, (int)data_len, ttl);
```
If compiled with `-Wall` you get a warning here about unmatched types (`%d` and `uint64_t`).
If ttl becomes `time_t` as mentioned above you can cast it to `unsigned int` and use `%u`.

gthess commented on cachedb/redis.c:
```diff
@@ -249,16 +250,37 @@ redis_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
+	else {
+		verbose(VERB_ALGO, "redis_store %s (%d bytes) with ttl %d", key, (int)data_len, ttl);
+		/* build command to set to a binary safe string */
+		n = snprintf(cmdbuf, sizeof(cmdbuf), "SET %s %%b EX %d", key, ttl);
```
Same as above.

gthess reviewed (changes requested):
Nice work! Although I have left some remarks that need to be addressed in the code before merging.

Some features that I would also like to see with this PR (ordered by difficulty/amount of work):
- Use of `SETEX` instead of `SET .. EX .. `. From the documentation SETEX is not deprecated and supported in earlier versions than SET EX.
- Configuration option for redis to explicitly allow this behavior or not. (Default would be no, in order to maintain backwards compatibility)
- Detect SETEX redis support when cachedb initiates and error out if not.

If you want to tackle them as well feel free! Otherwise let me know and I will add them with this PR.

Talkabout commented:
HI @gthess ,

thanks for the review, I will tackle the requested changes on my end to get some experience with changing unbound code.

Bye

gthess commented on cachedb/redis.c:
```diff
@@ -249,16 +250,37 @@ redis_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
+		size = 4+(CACHEDB_HASHSIZE/8)*2+3+4+sizeof(uint64_t)+1;
+	}
+
+	char cmdbuf[size]; /* "SET " + key + " %b EX " + ttl */
```
This should move uptop with the rest of the declarations in this code block (function) to make the code more portable (pre-std99 complain about mixed declarations and code).

Talkabout commented:
Hi @gthess ,

I have now added the changes. Creating a new option is quite a complex task... ;)

Now my git shows me the following files as changed:

```
cachedb/cachedb.c
cachedb/cachedb.h
cachedb/redis.c
util/config_file.c
util/config_file.h
util/configlexer.c
util/configlexer.lex
util/configparser.c
util/configparser.h
util/configparser.y
```
some of them are auto generated so the question is, if I should include all of them in my commit?

Talkabout committed:
added option 'redis-set-ttl' to define whether ttl should be added to redis records
added check for redis command 'setex' when initializing redis connection
updated documentation
minor improvements to previous changes

Talkabout commented:
I committed all changed files, please merge only the ones you think are necessary.

Thanks!

gthess commented:
Thanks, from a first look everything seems good and in place!
But I do have some remarks. Will resolve my previous remarks and offer a new review later today.

Talkabout commented:
Great! Thanks!

gthess commented on cachedb/redis.c:
```diff
@@ -114,6 +118,24 @@ redis_init(struct module_env* env, struct cachedb_env* cachedb_env)
+
+	if (env->cfg->redis_set_ttl) {
+		/** check if setex command is supported */
+		rep = redis_command(env, cachedb_env, "SETEX __UNBOUND_REDIS_CHECK__ 1 none", NULL, 0);
```
When `rep != NULL` you need to free it with `freeReplyObject()` before returning, otherwise there will be memory leaks.

gthess commented on cachedb/redis.c:
```diff
@@ -114,6 +118,24 @@ redis_init(struct module_env* env, struct cachedb_env* cachedb_env)
+		switch (rep->type) {
+		case REDIS_REPLY_ERROR:
+			/** init failed, setex command not supported */
+			log_err("redis_init: failed to init redis, setex command not supported?");
```
You could change both this and the above log to
>redis_init: failed to init redis, the redis-set-timeout option requires the SETEX command (redis >= 2.0.0)

to make debugging easier.
The first error log could stay the same but current behavior is to not error out if the connection to redis cannot be made; future connection could still happen. That way we make it clearer that we fail to start because we can't determine SETEX support.

gthess commented on cachedb/redis.c:
```diff
@@ -249,16 +271,27 @@ redis_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
-	char cmdbuf[4+(CACHEDB_HASHSIZE/8)*2+3+1]; /* "SET " + key + " %b" */
 	int n;
+	int set_ttl = (int)(env->cfg->redis_set_ttl && (!env->cfg->serve_expired || env->cfg->serve_expired_ttl > 0));
+	char cmdbuf[4+(CACHEDB_HASHSIZE/8)*2+3+((2+11)*set_ttl)+1]; /* "SETEX " + key + " " + ttl + " %b" */ 
```
There are 2 commands that we need to support (I use underscores for spaces):
- `SET_<KEY>_%b\0`
  4 + (CACHEDB_HASHSIZE/8)*2 + 3 +1
- `SETEX_<KEY>_<TTL>_%b\0`
  6 + (CACHEDB_HASHSIZE/8)*2 + 11 + 3 +1

I believe that the correct value to use here is the second one (longer one) unless I miss something.

Talkabout replied:
`((2+11)*set_ttl)`

this basically added the required additional bytes to the string size. As it is obviously easy to miss do you think I should write an if/else block?

gthess replied:
To clarify my comment, your calculation is correct but it is better/clearer to use the longer one (in case we want to expand this further). Having enough space for the command is what we care most.

gthess replied:
.. and you can put as a comment above which commands are currently supported.

Talkabout replied:
will do

gthess commented on cachedb/redis.c:
```diff
@@ -249,16 +271,27 @@ redis_lookup(struct module_env* env, struct cachedb_env* cachedb_env,
+		/* build command to set to a binary safe string */
+		n = snprintf(cmdbuf, sizeof(cmdbuf), "SET %s %%b", key);
+	}
+	else {
```
Formatting: `} else {` to be consistent with the rest of the code.

gthess commented on cachedb/redis.c:
```diff
@@ -114,6 +118,24 @@ redis_init(struct module_env* env, struct cachedb_env* cachedb_env)
+			/** init failed, no response from redis server*/
+			log_err("redis_init: failed to init redis, could not connect to server");
+			return 0;
+		switch (rep->type) {
```
From the [documentation](https://redis.io/commands/setex) SETEX returns only a `REDIS_REPLY_STRING`. I would break on that (what you currently do with default) and error out on default instead.

Talkabout replied:
I was not able to apply this change. I tested it but somehow redis (keydb in my case, which is 100% compatible with redis) returns "5" as response type, which is not defined in hiredis library. That is why I kept it the way it was before. Maybe you have an idea why it behaves that way.

gthess replied:
My bad, 5 (REDIS_REPLY_STATUS) is the correct value. The documentation was mentioning "simple string reply" and I hastily thought REDIS_REPLY_STRING was the one.

Talkabout replied:
Very good hint! I have checked the "hiredis" code but was not able to find the constant with value "5". I will add the change. Thanks!

gthess replied:
I found it [here](https://github.com/redis/hiredis/blob/master/read.h#L55)

Talkabout replied:
Ok, again missed something "obvious"... sorry!

gthess commented on doc/unbound.conf.5.in:
```diff
@@ -2127,6 +2127,13 @@ If this timeout expires Unbound closes the connection, treats it as
+that redis can evict keys that have expired automatically. If unbound
+is configured to serve expired entries and there is no expired ttl set,
+this option is internally reverted to "no".
+This option defaults to no.
```
I would mention that support for SETEX is required for this to work (redis >= 2.0.0).

gthess reviewed (changes requested):
Nice almost done!
Apart from the remarks one editorial nit: I would use redis-expire-records instead of redis-set-ttl to avoid any confusion with the actual DNS TTL.

Talkabout committed:
implemented review feedback
renamed option from 'redis-set-ttl' to 'redis-expire-records'

Talkabout committed:
changed init logic of redis backend as per review request

gthess commented:
Thanks for sticking with this!

Talkabout commented:
Thanks for your help and guidance!
